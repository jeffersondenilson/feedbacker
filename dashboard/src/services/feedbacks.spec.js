import mockAxios from 'axios'
import FeedbacksService from './feedbacks'
import feedbacks from './__mocks__/feedbacks'

jest.mock('axios')

describe('FeedbacksService', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should return feedbacks within default pagination', async () => {
    const expectedPagination = {
      limit: 5,
      offset: 0,
      total: feedbacks.length
    }
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { limit, offset } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks.slice(offset, limit),
          pagination: { limit, offset, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll()

    expect(response.data).toHaveProperty('results')
    expect(response.data).toHaveProperty('pagination')
    expect(response.data.pagination).toStrictEqual(expectedPagination)
    expect(response).toMatchSnapshot()
  })

  it('should return feedbacks within offset = 2 limit = 5', async () => {
    const expectedPagination = {
      limit: 5,
      offset: 2,
      total: feedbacks.length
    }
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { limit, offset } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks
            .slice(offset, feedbacks.length)
            .slice(0, limit),
          pagination: { limit, offset, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      limit: 5,
      offset: 2
    })

    expect(response.data).toHaveProperty('results')
    expect(response.data).toHaveProperty('pagination')
    expect(response.data.pagination).toStrictEqual(expectedPagination)
    expect(response).toMatchSnapshot()
  })

  it('should return feedbacks within offset = 5 limit = 5', async () => {
    const expectedPagination = {
      limit: 5,
      offset: 5,
      total: feedbacks.length
    }
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { limit, offset } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks
            .slice(offset, feedbacks.length)
            .slice(0, limit),
          pagination: { limit, offset, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      limit: 5,
      offset: 5
    })

    expect(response.data).toHaveProperty('results')
    expect(response.data).toHaveProperty('pagination')
    expect(response.data.pagination).toStrictEqual(expectedPagination)
    expect(response).toMatchSnapshot()
  })

  it('should return feedbacks within offset = 3 limit = 2', async () => {
    const expectedPagination = {
      limit: 2,
      offset: 2,
      total: feedbacks.length
    }
    mockAxios.get.mockImplementationOnce((url, config) => {
      let { limit, offset } = config.params

      if (offset > limit) {
        offset = limit
      }

      return Promise.resolve({
        data: {
          results: feedbacks
            .slice(offset, feedbacks.length)
            .slice(0, limit),
          pagination: { limit, offset, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      limit: 2,
      offset: 3
    })

    expect(response.data).toHaveProperty('results')
    expect(response.data).toHaveProperty('pagination')
    expect(response.data.pagination).toStrictEqual(expectedPagination)
    expect(response).toMatchSnapshot()
  })

  it('should return feedbacks within offset = 0 limit = 11', async () => {
    const expectedPagination = {
      limit: 5,
      offset: 0,
      total: feedbacks.length
    }
    mockAxios.get.mockImplementationOnce((url, config) => {
      let { limit, offset } = config.params

      if (limit > 10) {
        limit = 5
      }

      return Promise.resolve({
        data: {
          results: feedbacks
            .slice(offset, feedbacks.length)
            .slice(0, limit),
          pagination: { limit, offset, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      limit: 11,
      offset: 0
    })

    expect(response.data).toHaveProperty('results')
    expect(response.data).toHaveProperty('pagination')
    expect(response.data.pagination).toStrictEqual(expectedPagination)
    expect(response).toMatchSnapshot()
  })

  it('should return all feedbacks of type ISSUE', async () => {
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { type } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks.filter(fb => fb.type === type.toUpperCase()),
          pagination: { limit: 5, offset: 0, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      type: 'issue'
    })

    expect(response.data).toHaveProperty('results')
    expect(response).toMatchSnapshot()
  })

  it('should return all feedbacks of type IDEA', async () => {
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { type } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks.filter(fb => fb.type === type.toUpperCase()),
          pagination: { limit: 5, offset: 0, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      type: 'idea'
    })

    expect(response.data).toHaveProperty('results')
    expect(response).toMatchSnapshot()
  })

  it('should return all feedbacks of type OTHER', async () => {
    mockAxios.get.mockImplementationOnce((url, config) => {
      const { type } = config.params
      return Promise.resolve({
        data: {
          results: feedbacks.filter(fb => fb.type === type.toUpperCase()),
          pagination: { limit: 5, offset: 0, total: feedbacks.length }
        }
      })
    })

    const response = await FeedbacksService(mockAxios).getAll({
      type: 'other'
    })

    expect(response.data).toHaveProperty('results')
    expect(response).toMatchSnapshot()
  })

  it('should return count of feedback types', async () => {
    const summary = { all: 7, issue: 3, idea: 3, other: 1 }
    mockAxios.get.mockImplementationOnce(() => {
      return Promise.resolve({ data: summary })
    })

    const response = await FeedbacksService(mockAxios).getSummary()

    expect(response).toMatchSnapshot({
      data: {
        all: expect.any(Number),
        issue: expect.any(Number),
        idea: expect.any(Number),
        other: expect.any(Number)
      }
    })
  })
})
