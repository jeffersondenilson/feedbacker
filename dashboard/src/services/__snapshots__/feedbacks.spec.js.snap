// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbacksService should return all feedbacks of type IDEA 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 0,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1600819200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1608681600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 1",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1598140800000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 2",
        "type": "IDEA",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return all feedbacks of type ISSUE 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 0,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1603411200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk 2",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Tava bom, agora parece que piorou",
        "type": "ISSUE",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return all feedbacks of type OTHER 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 0,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1605225600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muito bom!",
        "type": "OTHER",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return count of feedback types 1`] = `
Object {
  "data": Object {
    "all": Any<Number>,
    "idea": Any<Number>,
    "issue": Any<Number>,
    "other": Any<Number>,
  },
}
`;

exports[`FeedbacksService should return feedbacks within default pagination 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 0,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1605225600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muito bom!",
        "type": "OTHER",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1603411200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1600819200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1608681600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 1",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1598140800000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 2",
        "type": "IDEA",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return feedbacks within offset = 0 limit = 11 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 0,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1605225600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muito bom!",
        "type": "OTHER",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1603411200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1600819200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1608681600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 1",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1598140800000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 2",
        "type": "IDEA",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return feedbacks within offset = 2 limit = 5 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 2,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1600819200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1608681600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 1",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1598140800000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 2",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk 2",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Tava bom, agora parece que piorou",
        "type": "ISSUE",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return feedbacks within offset = 3 limit = 2 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 2,
      "offset": 2,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1600819200000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo",
        "type": "IDEA",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1608681600000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Podia ter um botão de solicitar demo 1",
        "type": "IDEA",
      },
    ],
  },
}
`;

exports[`FeedbacksService should return feedbacks within offset = 5 limit = 5 1`] = `
Object {
  "data": Object {
    "pagination": Object {
      "limit": 5,
      "offset": 5,
      "total": 7,
    },
    "results": Array [
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Muitos erros slkkkkkkk 2",
        "type": "ISSUE",
      },
      Object {
        "apiKey": "fcd5015c-10d3-4e9c-b395-ec7ed8850165",
        "createdAt": 1590192000000,
        "device": "Chrome 85.0, macOS 10.14",
        "fingerprint": "490135491",
        "id": "eab759f8-f238-4ff9-ae91-ee1558982329",
        "page": "https://feedbacker.com/pricing",
        "text": "Tava bom, agora parece que piorou",
        "type": "ISSUE",
      },
    ],
  },
}
`;
